internal/providers/dockerhub/dockerhub.go

```diff
@@ -0,0 +1,151 @@
+	db.ProviderTypeOci,
+}
+
+// AuthorizationFlows is the list of authorization flows that the DockerHub provider supports
+var AuthorizationFlows = []db.AuthorizationFlow{
+	db.AuthorizationFlowUserInput,
+}
+
+// dockerHubImageLister is the struct that contains the Docker Hub specific operations
+type dockerHubImageLister struct {
+	cred      provifv1.OAuth2TokenCredential
+	cli       *http.Client
+	namespace string
+	target    *url.URL
+	cfg       *minderv1.DockerHubProviderConfig
+}
+
+// Ensure that the Docker Hub client implements the ImageLister interface
+var _ provifv1.ImageLister = (*dockerHubImageLister)(nil)
+
+// New creates a new Docker Hub client
+func New(cred provifv1.OAuth2TokenCredential, cfg *minderv1.DockerHubProviderConfig) (*dockerHubImageLister, error) {
+	cli := oauth2.NewClient(context.Background(), cred.GetAsOAuth2TokenSource())
+
+	u, err := url.Parse("https://hub.docker.com/v2/repositories")
+	if err != nil {
+		return nil, fmt.Errorf("error parsing base URL: %w", err)
+	}
+
+	ns := cfg.GetNamespace()
+	t := u.JoinPath(ns)
+
+	return &dockerHubImageLister{
+		namespace: ns,
+		cred:      cred,
+		cli:       cli,
+		target:    t,
+		cfg:       cfg,
+	}, nil
+}
+
+// ParseV1Config parses the raw config into a DockerHubProviderConfig struct
+func ParseV1Config(rawCfg json.RawMessage) (*minderv1.DockerHubProviderConfig, error) {
+	type wrapper struct {
+		DockerHub *minderv1.DockerHubProviderConfig `json:"dockerhub" yaml:"dockerhub" mapstructure:"dockerhub" validate:"required"`
+	}
+
+	var w wrapper
+	if err := provifv1.ParseAndValidate(rawCfg, &w); err != nil {
+		return nil, err
+	}
+
+	// Validate the config according to the protobuf validation rules.
+	if err := w.DockerHub.Validate(); err != nil {
+		return nil, fmt.Errorf("error validating GitHubOAuth v1 provider config: %w", err)
+	}
+
+	return w.DockerHub, nil
+}
+
+func (d *dockerHubImageLister) GetNamespaceURL() string {
+	return d.target.String()
+}
+
+// CanImplement returns true if the provider can implement the specified trait
+func (_ *dockerHubImageLister) CanImplement(trait minderv1.ProviderType) bool {
+	return trait == minderv1.ProviderType_PROVIDER_TYPE_DOCKERHUB
+}
+
+// ListImages lists the containers in the Docker Hub
+func (d *dockerHubImageLister) ListImages(ctx context.Context) ([]string, error) {
+	req, err := http.NewRequest("GET", d.target.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, err := d.cli.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, fmt.Errorf("error making request: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusUnauthorized {
+			return nil, fmt.Errorf("unauthorized: %s", resp.Status)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, errors.New("not found")
+		}
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	// read body
+	defer resp.Body.Close()
```

Comment:
please move the defer above or else you might leak the body during error handling.